### DISCUSSION TITLE
Steering issue of Mushr racecar

### DISCUSSION THREAD

**Discussion by Lilly-yang** [prl-mushr/mushr#25] [Hardware]
Hi,

I am trying to use a motor controller to control the Mushr racecar, but I met some problems when I want to turn right/left. Here are the details:

- When we push the left joystick forward or backward, the wheels will scroll forward or backward, and no message will pop up in the terminal window.

- When we push the right joystick to the left, the wheels didn't turn to the left, and an error pops up in the terminal window. When we push the right joystick to the left while pushing the left joystick forward or backward at the same time, the wheel will not stop.

- When we push the right joystick to the right, the wheels didn't turn to the right, but we can hear some sounds, and sometimes an error message pops up in the terminal window. When we push the right joystick to the right while pushing the left joystick forward or backward at the same time, the wheel will stop.

Here is the screenshot of the error message.
![image](https://user-images.githubusercontent.com/39558023/118826667-f85a6180-b8bb-11eb-8a24-5f76ad570866.png)

And there is a [video](https://drive.google.com/file/d/1XrdnevbrrCzovlZkMaSMw-GtxgnJ6fh-/view?usp=sharing) showing what happened.

And we also tried to control the racecar with the keyboard via directly connect VESC to VESC-Tool. And we met a different problem:

- The "up" and "down" keys (reversely) control the tires to "backward" and "forward" rolling;
- The 'left' and 'right' keys also control the tires to 'forward' and 'backward' rolling but way faster.
- If we manually turn the wheels to the right or left, and push the 'up' and 'down' keys again, the tires sometimes get stuck. But no problem with the 'left' and 'right' keys I think maybe because their acceleration command provides more energy to the motor and the tires can still rotate smoothly even in corners.

And there is also a [video](https://drive.google.com/file/d/1hzXNfGp1ETZomOMpiU3aGQGSZYZ30MJA/view?usp=sharing) showing what happened.

We also tried some method to check where is the problem but also failed:
![image](https://user-images.githubusercontent.com/39558023/118832481-f050f080-b8c0-11eb-96d6-c3683136f8bd.png)
and no responses for this command: `rostopic pub -r 10 /commands/servo/position std_msgs/Float64 0.75`

More information:

1. We configured the software of Mushr racecar followed this [tutorial](https://mushr.io/hardware/build_instructions/#software-setup).
2. The servo we use is [servo 2Kg, Metal Arm](https://www.amazon.com/gp/product/B01MU7TQV8)
3. we used VESC-Tool to upload and write mushr_vesc_sensesc_sensorless_config.xml, which updated 6 days ago.
4. we made some changes just to make sure that the parameters are the same as in the tutorial.
![image](https://user-images.githubusercontent.com/39558023/118832113-a1a35680-b8c0-11eb-9345-f012932d1b3d.png)

Here are some information read by VESC-Tool:
![image](https://user-images.githubusercontent.com/39558023/118831109-c814c200-b8bf-11eb-810a-7dad1f232527.png)
![image](https://user-images.githubusercontent.com/39558023/118831304-f4304300-b8bf-11eb-9638-5e46457ff45c.png)
![image](https://user-images.githubusercontent.com/39558023/118831371-04482280-b8c0-11eb-98f0-637ac0e58b16.png)
![image](https://user-images.githubusercontent.com/39558023/118831475-1924b600-b8c0-11eb-9d76-cae661143127.png)
![image](https://user-images.githubusercontent.com/39558023/118831518-2346b480-b8c0-11eb-9d77-76eb30dac3de.png)
![image](https://user-images.githubusercontent.com/39558023/118831624-348fc100-b8c0-11eb-9d7f-75469ef1aa52.png)
![image](https://user-images.githubusercontent.com/39558023/118831707-496c5480-b8c0-11eb-9c49-0ae3646090c4.png)

Please feel free to let me know if you need more information. Thanks in advance!

**Comment by palanc** [comment]
Hi @Lilly-yang - is it possible that you uploaded the wrong firmware? From one of your screenshots, it looks like you may have uploaded the hw_410_411_412/**VESC_default.bin** file, but what you actually want to upload is hw_410_411_412/**VESC_servoout.bin**. If you upload the correct firmware, does it work as expected?

If not, could you change the sharing permissions on the videos you posted? I currently cannot view them because I don't have access.

Also, if the wheels are turning the opposite way of what is expected (e.g. wheels are moving backwards when tilting left joystick forwards), check the build tutorial again to see if the brushless motor wire are connected to the VESC wires in the correct order. Alternatively, you can swap any pair of wires and it should undo the reversal.

**Reply by Lilly-yang** [reply]
Thank you for your quick and considerable reply.

Actually, I have already solved this problem yesterday. Somebody already answered this question in the [forum](https://spectrum.chat/mushr/hardware/couldnt-update-vesc-firmware-with-bldc-tool~108ef6b9-f37c-4ca4-b724-7b4c60e1fba1).
![image](https://user-images.githubusercontent.com/39558023/118981782-058a5580-b97b-11eb-912b-eca181ae5f81.png)

I upload the new version of `VESC_default.bin` because the `VESC_servoout.bin`, which is no longer offered in the latest version of [firmware](https://github.com/vedderb/vesc_tool/tree/master/res/firmwares/410_o_411_o_412) for VESC, didn't match the current version of the VESC tool and I couldn't use it to upload the `mushr_vesc_sensorless_config.xml`. And I didn't use the BLDC tool because it no longer supports Ubuntu 18.

The problems caused by version iteration are kinda complicated. Fortunately, it has been resolved (at least for now). ：）

**Reply by schmittlema** [reply]
@planc509 could we update the build instructions to make sure the vesc version is correct? Or do we need to update our vesc firmware to make this compatible with the latest?

**Comment by palanc** [comment]
@schmittlema I think we need some clarification first

@Lilly-yang I'm glad that you were able to fix the problem, could you tell us more about the solution so we can better understand what was wrong? Specifically, are you saying that just uploading VESC_default.bin solved the issue? Or are there other steps that you took to fix it?

**Reply by Lilly-yang** [reply]
Oh, I may not express it clearly.

I tried the latest `VESC_default.bin` only because the new version of the VESC tool does not support writing `mushr_vesc_sensorless_config.xml` to the old version of the VESC firmware (VESC_servoout.bin was updated 5 years ago), and my computer system is Ubuntu 18 and cannot use the BLDC tool.
(I think it's may because of the system. I also can’t use the latest version of the VESC tool on Ubuntu 16.)

In the end, according to the answer I found,  I back to used the BLDC tool to write `mushr_vesc_sensorless_config.xml` just in the Ubuntu 16 system. Actually, it is the same as the tutorial except on two different systems: ​
1. Use the latest version of the VESC tool (on Ubuntu 18) to upload VESC_servoout.bin;
2. Use the BLDC tool (in Ubuntu 16) to upload the latest mushr_vesc_sensorless_config.xml.

So the current problem is only the support of the BLDC tool and VESC tool to different systems. But considering that the BLDC tool has been replaced by the VESC tool for a long time, it is better to update the tutorial (which you are updating now). 

BTW, I also found some tiny mistake on the Global Planner tutorial such as the Misspell
![image](https://user-images.githubusercontent.com/39558023/119058934-9b4fd000-b9cf-11eb-8784-3d64abe6f633.png)
and possibly incomplete command because I must specify the package name when I execute it 
![image](https://user-images.githubusercontent.com/39558023/119059185-19ac7200-b9d0-11eb-9d05-fae93e5faed1.png)
![image](https://user-images.githubusercontent.com/39558023/119059020-c9cdab00-b9cf-11eb-87ee-0eb55b36ac66.png)

Hope this information is useful. :)

**Reply by schmittlema** [reply]
Very interesting, thanks for the detail. So if I'm understanding correctly it seems we need to update the vesc firmware so the vesc tool can write the xml to a new firmware. And we also need to replace BLDC tool instructions/use with vesc tool? 

In regards to that tutorial, thanks for the heads up! We are writing a new global planner tutorial soon.

**Reply by Lilly-yang** [reply]
Sorry for my late reply. It is of course the best to keep the software and hardware as the latest adapted version and I think you will definitely update them when the next major hardware/software upgrade. But according to the current situation, as long as the BDLC and VESC tools can be used normally, there is no problem with the rest according to the tutorial.

**Reply by berktepebag** [reply]
> Very interesting, thanks for the detail. So if I'm understanding correctly it seems we need to update the vesc firmware so the vesc tool can write the xml to a new firmware. And we also need to replace BLDC tool instructions/use with vesc tool?
> 
> In regards to that tutorial, thanks for the heads up! We are writing a new global planner tutorial soon.

Dear @schmittlema , I went with VESC 6 to be more future-proof, do you think it is possible to also consider VESC-6 as part of the tutorials?
